### PR TITLE
feat(delayedack): eibc demand order on ack error

### PR DESCRIPTION
**Pull request by danwt**
## Description

----

Closes https://github.com/dymensionxyz/dymension/issues/728
Partially addresses https://github.com/dymensionxyz/dymension/issues/540

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [x]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x]  Targeted PR against the correct branch
- [x]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x]  Linked to the GitHub issue with discussion and accepted design
- [x]  Targets only one GitHub issue
- [x]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [x]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Updated the scripts for local run, e.g genesis_config_commands.sh if the PR changes parameters
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----;

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

---;

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

### CODE REVIEW

File: x/delayedack/ibc_middleware.go [hunk 1]
```diff
@@ -177,19 +112,33 @@ func (im IBCMiddleware) OnAcknowledgementPacket(
...
+		logger.Error("Unmarshal acknowledgement", "err", err)
+		return errorsmod.Wrapf(sdkerrors.ErrUnknownRequest, "unmarshal ICS-20 transfer packet acknowledgement: %v", err)
+	}
+
+	isErrAck := false
+
+	switch ack.Response.(type) {
+	case *channeltypes.Acknowledgement_Error:
+		isErrAck = true
+		// TODO(danwt): emit event? that's what the ibc transfer module does
```
[the remark this answers is not in the document]
Comment (danwt) [reply x/delayedack/ibc_middleware.go:130]: good point 👍

File: x/delayedack/ibc_middleware.go [hunk 2]
```diff
@@ -177,19 +112,33 @@ func (im IBCMiddleware) OnAcknowledgementPacket(
...
+	var ack channeltypes.Acknowledgement
+	if err := types.ModuleCdc.UnmarshalJSON(acknowledgement, &ack); err != nil {
+		logger.Error("Unmarshal acknowledgement", "err", err)
+		return errorsmod.Wrapf(sdkerrors.ErrUnknownRequest, "unmarshal ICS-20 transfer packet acknowledgement: %v", err)
+	}
+
+	isErrAck := false
+
+	switch ack.Response.(type) {
+	case *channeltypes.Acknowledgement_Error:
```
[the remark this answers is not in the document]
Comment (danwt) [reply x/delayedack/ibc_middleware.go:128]: fixed

File: x/delayedack/eibc.go [hunk 3]
```diff
@@ -43,29 +43,51 @@ func (im IBCMiddleware) eIBCDemandOrderHandler(ctx sdk.Context, rollappPacket co
 		// Unmarshal the packet metadata from the memo
 		err = json.Unmarshal([]byte(data.Memo), packetMetaData)
 		if err != nil {
-			logger.Error("error parsing packet metadata from memo", "error", err)
+			logger.Error("parsing packet metadata from memo", "error", err)
 			return nil
 		}
-		// If the rollapp packet is of type ON_TIMEOUT, the function will calculate the fee and create a demand order from the packet data.
-	} else if rollappPacket.Type == commontypes.RollappPacket_ON_TIMEOUT {
+	} else
```
[the remark this answers is not in the document]
Comment (danwt) [reply x/delayedack/eibc.go:49]: fixed

File: x/delayedack/eibc.go [hunk 4]
```diff
@@ -99,47 +118,51 @@ func (im IBCMiddleware) createDemandOrderFromIBCPacket(fungibleTokenPacketData t
...
+	// TODO(danwt): explain this calculation
 	demandOrderPrice := amountInt.Sub(feeInt).String()
-	// Get the denom for the demand order. If it's a timeout packet
-	// than its simply the denom we tried to send. If it's a receive packet
-	// than it's the IBC denom we've got.
+
 	var demandOrderDenom string
 	var demandOrderRecipient string
+	// Get the denom for the demand order
 	if rollappPacket.Type == commontypes.RollappPacket_ON_TIMEOUT {
```
Comment (mtsitrin) [inline x/delayedack/eibc.go:144]: switch case?

File: x/delayedack/eibc.go [hunk 5]
```diff
@@ -99,47 +118,51 @@ func (im IBCMiddleware) createDemandOrderFromIBCPacket(fungibleTokenPacketData t
...
 	if !ok || !feeInt.IsPositive() {
-		return nil, fmt.Errorf("Failed to convert fee to positive integer, %s", fee)
+		return nil, fmt.Errorf("convert fee to positive integer: %s", fee)
 	}
 	if amountInt.LT(feeInt) {
-		return nil, fmt.Errorf("Fee cannot be larger than amount")
+		return nil, fmt.Errorf("fee cannot be larger than amount")
 	}
+
+	// TODO(danwt): explain this calculation
```
Comment (mtsitrin) [inline x/delayedack/eibc.go:138]: gonna handle it? or remove

File: x/eibc/types/params.go [hunk 6]
```diff
@@ -15,11 +15,14 @@ var (
...
 	// KeyTimeoutFee is the key for the timeout fee
 	KeyTimeoutFee = []byte("TimeoutFee")
+	// KeyErrAckFee is the key for the error acknowledgement fee
+	KeyErrAckFee = []byte("ErrAckFee")
 )
 
 const (
 	defaultEpochIdentifier = "hour"
 	defaultTimeoutFee      = "0.0015"
+	defaultErrAckFee       = "0.0015" // TODO(danwt): check
```
Comment (mtsitrin) [inline x/eibc/types/params.go:25]: remove todo

File: x/delayedack/eibc.go (hunk 5 above)
Comment (danwt) [reply x/delayedack/eibc.go:138]: fixed

File: x/eibc/types/params.go (hunk 6 above)
Comment (danwt) [reply x/eibc/types/params.go:25]: fixed

File: x/delayedack/eibc.go (hunk 4 above)
Comment (danwt) [reply x/delayedack/eibc.go:144]: fixed

File: x/delayedack/eibc.go (hunk 5 above)
Comment (danwt) [reply x/delayedack/eibc.go]: added a explanation

**Review by mtsitrin — APPROVED** [review]

File: x/delayedack/eibc.go [hunk 7]
```diff
@@ -99,47 +118,64 @@ func (im IBCMiddleware) createDemandOrderFromIBCPacket(fungibleTokenPacketData t
...
-		demandOrderDenom = fungibleTokenPacketData.Denom
-		demandOrderRecipient = fungibleTokenPacketData.Sender
-	} else if rollappPacket.Type == commontypes.RollappPacket_ON_RECV {
+	// Get the denom for the demand order
+	switch rollappPacket.Type {
+	case commontypes.RollappPacket_ON_TIMEOUT:
+		fallthrough
+	case commontypes.RollappPacket_ON_ACK:
+		demandOrderDenom = fungibleTokenPacketData.Denom      // It's what we tried to send
+		demandOrderRecipient = fungibleTokenPacketData.Sender // and to who we tried to send it to
```
Comment (omritoptix) [inline x/delayedack/eibc.go:162]: I believe the comment here is wrong. it's not who we tried to send it to, but the address which sent it from

File: x/delayedack/eibc.go [hunk 8]
```diff
@@ -99,47 +118,64 @@ func (im IBCMiddleware) createDemandOrderFromIBCPacket(fungibleTokenPacketData t
...
+	case commontypes.RollappPacket_ON_RECV:
 		demandOrderDenom = im.getEIBCTransferDenom(*rollappPacket.Packet, fungibleTokenPacketData)
-		demandOrderRecipient = fungibleTokenPacketData.Receiver
+		demandOrderRecipient = fungibleTokenPacketData.Receiver // we're the one getting it, just faster
```
Comment (omritoptix) [inline x/delayedack/eibc.go:165]: this is who we tried sending it to (but from the rollapp not from the hub)

File: x/delayedack/eibc.go (hunk 7 above)
Comment (danwt) [reply x/delayedack/eibc.go:162]: ah yep good spot

File: x/delayedack/eibc.go [hunk 9]
```diff
@@ -43,29 +44,47 @@ func (im IBCMiddleware) eIBCDemandOrderHandler(ctx sdk.Context, rollappPacket co
...
+			logger.Debug("ErrAck fee is zero, skipping demand order creation")
+			return nil
+		}
+		fee := amountDec.Mul(feeMultiplier).TruncateInt().String()
 		packetMetaData = &types.PacketMetadata{
 			EIBC: &types.EIBCMetadata{
 				Fee: fee,
 			},
 		}
 	}
```
[the remark this answers is not in the document]
Comment (danwt) [reply x/delayedack/eibc.go:86]: good spot

**Review by zale144 — APPROVED** [review]
lgtm

**Review by mtsitrin — APPROVED** [review]

**Pull request merged**